File: RailgunVB/Commands/Music/MusicReset.vb
Imports AudioChord
Imports Discord
Imports Discord.Commands
Imports Microsoft.Extensions.DependencyInjection
Imports MongoDB.Bson
Imports RailgunVB.Core
Imports RailgunVB.Core.Configuration
Imports RailgunVB.Core.Managers
Imports RailgunVB.Core.Preconditions
Imports TreeDiagram
Imports TreeDiagram.Models.Server

Namespace Commands.Music
    Partial Public Class Music
        
        <Group("reset"), UserPerms(GuildPermission.ManageGuild)>
        Public Class MusicReset
            Inherits SystemBase
            
            Private ReadOnly _config As MasterConfig
            Private ReadOnly _playerManager As PlayerManager
            Private ReadOnly _musicService AS MusicService
            Private _full As Boolean = False

            Public Sub New(config As MasterConfig, playerManager As PlayerManager, musicService As MusicService)
                _config = config
                _playerManager = playerManager
                _musicService = musicService
            End Sub
            
            <Command("stream")>
            Public Async Function StreamAsync() As Task
                If Not (_playerManager.IsCreated(Context.Guild.Id))
                    If Not (_full) Then await ReplyAsync("I'm not streaming any music at this time.")
                    Return
                End If
                
                _playerManager.DisconnectPlayer(Context.Guild.Id)
                
                If Not (_full) Then await ReplyAsync($"Music stream has been reset! Use {Format.Code(
                    $"{_config.DiscordConfig.Prefix}music join")} to create a new music stream.")
            End Function
            
            <Command("playlist")>
            Public Async Function PlaylistAsync() As Task
                Dim data As ServerMusic = Await Context.Database.ServerMusics.GetAsync(Context.Guild.Id)
                    
                If data Is Nothing OrElse data.PlaylistId = ObjectId.Empty AndAlso Not (_full)
                    Await ReplyAsync("Server playlist is already empty.")
                    Return
                End If
                
                Await StreamAsync()
                Await _musicService.Playlist.DeleteAsync(data.PlaylistId)
                
                data.PlaylistId = ObjectId.Empty
                
                If Not (_full)
                    Await ReplyAsync("Server playlist is now empty.")
                End If
            End Function
            
            <Command("full")>
            Public Async Function FullAsync() As Task
                _full = True
                
                Await PlaylistAsync()
                
                Dim data As ServerMusic = Await Context.Database.ServerMusics.GetAsync(Context.Guild.Id)
                
                If data Is Nothing
                    await ReplyAsync("Music has no data to reset.")
                    Return
                End If
                
                Context.Database.ServerMusics.Remove(data)
                await ReplyAsync("Music settings & playlist has been reset.")
            End Function
            
        End Class
        
    End Class
    
End NameSpace

File: Railgun/Commands/Music/MusicReset.cs
using System.Threading.Tasks;
using AudioChord;
using Discord;
using Finite.Commands;
using MongoDB.Bson;
using Railgun.Core;
using Railgun.Core.Attributes;
using Railgun.Core.Configuration;
using Railgun.Music;
using TreeDiagram;
using TreeDiagram.Models;
using TreeDiagram.Models.Server;

namespace Railgun.Commands.Music
{
	public partial class Music
	{
		[Alias("reset"), UserPerms(GuildPermission.ManageGuild)]
		public class MusicReset : SystemBase
		{
			private readonly MasterConfig _config;
			private readonly PlayerController _playerController;
			private readonly MusicService _musicService;
			private bool _full;

			public MusicReset(MasterConfig config, PlayerController playerController, MusicService musicService)
			{
				_config = config;
				_playerController = playerController;
				_musicService = musicService;
			}

			private ServerMusic GetData(ulong guildId, bool create = false)
			{
				ServerProfile data;

				if (create)
					data = Context.Database.ServerProfiles.GetOrCreateData(guildId);
				else {
					data = Context.Database.ServerProfiles.GetData(guildId);

					if (data == null) 
						return null;
				}

				if (data.Music == null)
					if (create)
						data.Music = new ServerMusic();
				
				return data.Music;
			}

			[Command("stream")]
			public Task StreamAsync()
			{
				if (_playerController.GetPlayer(Context.Guild.Id) == null)
					if (!_full) return ReplyAsync("I'm not streaming any music at this time.");

				_playerController.DisconnectPlayer(Context.Guild.Id);
				if (!_full) return ReplyAsync($"Music stream has been reset! Use {Format.Code($"{_config.DiscordConfig.Prefix}music join")} to create a new music stream.");
				return Task.CompletedTask;
			}

			[Command("playlist")]
			public async Task PlaylistAsync()
			{
				var data = GetData(Context.Guild.Id);

				if (data == null || data.PlaylistId == ObjectId.Empty && !_full) {
					await ReplyAsync("Server playlist is already empty.");
					return;
				}

				await StreamAsync();
				await SystemUtilities.DeletePlaylistAsync(_musicService, data.PlaylistId);

				data.PlaylistId = ObjectId.Empty;

				if (!_full) await ReplyAsync("Server playlist is now empty.");
			}

			[Command("full")]
			public async Task FullAsync()
			{
				_full = true;

				await PlaylistAsync();

				var data = Context.Database.ServerProfiles.GetData(Context.Guild.Id);

				if (data == null || data.Music == null) {
					await ReplyAsync("Music has no data to reset.");
					return;
				}

				data.Music = null;
				await ReplyAsync("Music settings & playlist has been reset.");
			}
		}
	}
}